persistence/pre_persistence/persistence.go

```diff
@@ -0,0 +1,306 @@
+func (m *PrePersistenceContext) NewSavePoint(bytes []byte) error {
+	index := len(m.DBs)
+	newDB := NewMemDB()
+	if index == 0 {
+		return fmt.Errorf("%s", "zero length mock persistence context")
+	}
+	src := m.DBs[index-1]
+	if err := CopyMemDB(src, newDB); err != nil {
+		return err
+	}
+	m.SavePoints[hex.EncodeToString(bytes)] = index
+	m.DBs = append(m.DBs, newDB)
+	return nil
+}
+
+// RollbackToSavePoint Rollback save point
+// Needed in the case of failed transactions during proposal or blocks during validation
+func (m *PrePersistenceContext) RollbackToSavePoint(bytes []byte) error {
+	rollbackIndex, ok := m.SavePoints[hex.EncodeToString(bytes)]
+	if !ok {
+		return fmt.Errorf("save point not found")
+	}
+	toDelete := make([]string, 0)
+	// rollback savepoints map
+	for key, i := range m.SavePoints {
+		if i > rollbackIndex {
+			toDelete = append(toDelete, key)
+		}
+	}
+	for _, key := range toDelete {
+		delete(m.SavePoints, key)
+	}
+	// rollback
+	m.DBs = m.DBs[:rollbackIndex]
+	return nil
+}
+
+// AppHash creates a unique hash based on the global state object
+// NOTE: AppHash is an inefficient, arbitrary, mock implementation that enables the functionality
+// TODO written for replacement, taking any and all better implementation suggestions - even if a temporary measure
+func (m *PrePersistenceContext) AppHash() ([]byte, error) {
+	result := make([]byte, 0)
+	index := len(m.DBs) - 1
+	db := m.DBs[index]
+	it := db.NewIterator(&util.Range{})
+	it.First()
+	for ; it.Valid(); it.Next() {
+		result = append(result, it.Value()...)
+		// chunk into 100000 byte segments
+		if len(result) >= 100000 {
+			result = crypto.SHA3Hash(result)
+		}
+	}
+	it.Release()
+	// potential for double hash here
+	return crypto.SHA3Hash(result), nil
+}
+
+// Reset to the first save point
+func (m *PrePersistenceContext) Reset() error {
+	return m.RollbackToSavePoint(FirstSavePointKey)
+}
+
+// Commit the KV pairs to the parent (commit) db
+func (m *PrePersistenceContext) Commit() error {
+	index := len(m.DBs) - 1
+	db := m.DBs[index]
+	it := db.NewIterator(&util.Range{})
+	it.First()
+	for ; it.Valid(); it.Next() {
+		if err := m.Parent.GetCommitDB().Put(HeightKey(m.Height, it.Key()), it.Value()); err != nil {
+			return err
+		}
+	}
+	it.Release()
+	m.Release()
+	parentIt := m.Parent.GetCommitDB().NewIterator(&util.Range{
+		Start: HeightKey(m.Height, nil),
+		Limit: PrefixEndBytes(HeightKey(m.Height, nil)),
+	})
+	parentIt.First()
+	m.Height = m.Height + 1
+	// copy over the entire last height
+	for ; parentIt.Valid(); parentIt.Next() {
+		newKey := HeightKey(m.Height, KeyFromHeightKey(parentIt.Key()))
+		if err := m.Parent.GetCommitDB().Put(newKey, parentIt.Value()); err != nil {
+			return err
+		}
+	}
+	parentIt.Release()
+	return nil
+}
+
+func (m *PrePersistenceContext) Release() {
+	m.SavePoints = nil
+	for _, db := range m.DBs {
+		db.Reset()
+	}
+	m.DBs = nil
+	return
```

Comment:
(_minor comment_) Would be great to see this also reflected on a flag to prevent double releasing/useless releasing and also to allow for more fine-tuned testing and usage.

Reply:
Not clear on the actionable here

Reply:
Never made it in time here. This will depend on how often do we reset, but if we reset once, we can do this:

```
if: already reset? return
else: iterate and release.
```

The persistence context can store that flag.